cli/src/main/java/org/adamalang/cli/runtime/Argument.java

```diff
@@ -0,0 +1,151 @@
+        for (int i = 0; i < args.length; i++) {
+            String arg = args[i];
+            if (arg.equals("--no-color")) {
+                arguments.put("--no-color", new ArgumentItem("--no-color", "", "Determines if the output should have color"));
+            } else if (arg.equals("--json")) {
+                arguments.put("--json", new ArgumentItem("--json", "", "Determines if the output should be in json format"));
+            }
+        }
+
+        // Validate everything here...
+        if (args.length  == 0) {
+            group = null;
+            return;
+        }
+        if (args.length >= 1) {
+            // Groups could be hashmap
+
+            if (Util.equalsMultiple(args[0], "help", "--help", "-h")) {
+               group = null;
+               return;
+            } else {
+                CliElement thisGroup = CliElement.Groups.get(args[0]);
+                if (thisGroup != null) {
+                    group = thisGroup;
+                    argStart = 1;
+                } else {
+                    System.out.println("Command '" + args[0] + "' is not a valid command. See 'adama.jar --help'");
+                    valid = false;
+                    return;
+                }
+
+            }
+        }
+
+        if (args.length >= 2) {
+            if (Util.equalsMultiple(args[1], "help", "--help", "-h")) {
+                command = null;
+                // If the group is a command.
+                if (group.Commands.isEmpty()) {
+                    Help.displayHelp(group.name);
+                    valid = false;
+                }
+                return;
+            } else {
+                // If group is command
+                if (!group.Commands.isEmpty()) {
+                    CliElement thisCommand = group.Commands.get(args[1]);
+                    if (thisCommand != null) {
+                        command = thisCommand;
+                        argStart = 2;
+                    } else {
+                        System.out.println("Command '" + args[1] + "' is not a valid command of '" + group.name + "'. See 'adama.jar " + group.name + " --help'");
+                        valid = false;
+                    }
+                }
+            }
+        }
+
+        if (argStart == 2 && argStart < args.length && Util.equalsMultiple(args[argStart], "--help", "-h")) {
+            Help.displayHelp(group.name, command.name);
+            valid = false;
+            return;
+        }
+
+        Map<String, ArgumentItem> loop;
+        if (command != null) {
+            loop = command.Arguments;
+        } else {
+            loop = group.Arguments;
+        }
+
+
+
+        boolean anyInvalid = false;
+        // Does not check for arguments that are not available...
+        // Could be bad, could be good, let's see.
+        outerloop:
+        for (Map.Entry<String, ArgumentItem> entry : loop.entrySet()) {
+            ArgumentItem val = entry.getValue().copy();
```

Comment:
If you make ArgumentItem immutable, then the below val.value will not work. However, you could make ArgumentItem a key type by implementing equals, hashcode, and compareTo (from Comparable<ArgumentItem> such that you can use an ArgumentItem inside a map. This would transform the arguments a bit, but it suggests that there is a split from ArgumentKey mapping to ArgumentValue.